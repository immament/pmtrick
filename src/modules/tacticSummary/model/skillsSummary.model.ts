import { PlayerPositionType } from '../../../common/model/player.model';

export interface SkillsSummary {
    main: number;
    secondary: number;
    tertiary: number;
    physical: number;
    gs: number;
    position: PlayerPositionType;
    rank?: number;
}

type SkillsGroups = 'main' | 'secondary' | 'physical' | 'tertiary';

export const skillsSummaryOrder: SkillsGroups[] = ['main', 'secondary', 'physical', 'tertiary'];
