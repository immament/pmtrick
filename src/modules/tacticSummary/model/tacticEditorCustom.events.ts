export type TacticEditorCustomEvents = 'PMT_tacticInit' | 'PMT_tacticSquadChanged' | 'PMT_advTacticChanged';
