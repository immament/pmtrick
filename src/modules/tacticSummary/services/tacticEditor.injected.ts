import { injectScript } from '@src/common/services/utils/injectScript';
/* eslint-disable no-console */
import { TacticEditorCustomEvents } from '@src/modules/tacticSummary/model/tacticEditorCustom.events';

type ExtendedWindow = typeof window & {
    changed: () => void;
    pmtOldChanged?: () => void;
    selectedpositions: string[];
};

const anyWindow = window as ExtendedWindow;

function listenSquandChanges(): void {
    anyWindow.pmtOldChanged = anyWindow.changed;
    anyWindow.changed = () => {
        console.log('my changed');
        anyWindow.pmtOldChanged && anyWindow.pmtOldChanged();
        notify('PMT_tacticSquadChanged', anyWindow.selectedpositions);
    };
}

function listenAdvanceTacticChanges() {
    // injected to use jquery
    injectScript(`$('#adv_options select').each(function(a, el) {
        $(el).selectmenu({
            change: function(a, e) {
                $(el).change(),
                console.log('selectmenu changed', a, e );
                pmtOldChanged();
                document.dispatchEvent(
                    new CustomEvent('PMT_advTacticChanged', {
                        detail: []
                    })
                );
            }
        })
      });`);
}

function notify<T>(eventName: TacticEditorCustomEvents, data: T) {
    document.dispatchEvent(
        new CustomEvent<T>(eventName, {
            detail: data,
        }),
    );
}

function prepare() {
    console.log('prepare.......');
    const manager = document.querySelector('#managerTabs');
    if (!manager) {
        console.log('prepare....... managerTabs not avaiable..');
        setTimeout(prepare, 1000);
        return;
    }
    const style = window.getComputedStyle(manager);
    if (style.display === 'none') {
        console.log('prepare....... managerTabs no visible yet ... ');
        setTimeout(prepare, 1000);
        return;
    }

    listenSquandChanges();
    listenAdvanceTacticChanges();

    notify('PMT_tacticInit', 'prepare');
}

prepare();
