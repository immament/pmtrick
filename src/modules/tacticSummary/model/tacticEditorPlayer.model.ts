import { StatsPositionKeys } from '@src/modules/tacticSummary/model/tacticStats.model';

import { Player } from '../../../common/model/player.model';

export interface TacticEditorPosition<T extends TacticEditorPlayer = TacticEditorPlayer> {
    position?: string;
    positionGroup?: StatsPositionKeys;
    id?: string;
    player?: T;
}

export interface TacticEditorData<T extends TacticEditorPlayer = TacticEditorPlayer> {
    positions: TacticEditorPosition<T>[];
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TacticEditorPlayer extends Player {}
