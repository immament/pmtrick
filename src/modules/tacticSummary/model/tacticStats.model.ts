import { PlayerPositionType } from '@src/common/model/player.model';
import { PlayerSkillsEnum, PlayerSkillsType } from '@src/common/model/playerSkills.model';

export enum StatsPosition {
    'all' = 'all',
    'g' = 'g',
    'd' = 'd',
    'm' = 'm',
    'a' = 'a',
}

const statsPositionKeyToPlayerPositionMap: Record<StatsPositionKeys, PlayerPositionType> = {
    g: 'G',
    d: 'D',
    m: 'M',
    a: 'F',
};

export function positionGroupToPlayerPosition(positionGroup: StatsPositionKeys): PlayerPositionType {
    return statsPositionKeyToPlayerPositionMap[positionGroup];
}

export type StatsPositionKeys = 'g' | 'd' | 'm' | 'a';
export const StatsPositionKeysArr = [...Object.keys(StatsPosition)] as StatsPosition[];

export type StatsKeys = PlayerSkillsType | 'gs';
export const StatsKeysArr = [...Object.keys(PlayerSkillsEnum), 'gs'] as StatsKeys[];

export type PositionStatRecord = { sum: number; avg?: number };

export type StatRecord = Record<StatsKeys, PositionStatRecord> & { count: number };

export type TacticStatsSums = Record<StatsPosition, StatRecord>;

export type AdvanceTacticTypes =
    | 'offside'
    | 'pressingLow'
    | 'pressingHigh'
    | 'counterAttack'
    | 'markingZonal'
    | 'markingMan'
    | 'highBalls'
    | 'oneOnOnes'
    | 'gkStand'
    | 'gkRush'
    | 'longShots'
    | 'firstTimeShots';
export type AdvanceTacticsValues = Record<AdvanceTacticTypes, (number | undefined)[]>;

export class TacticStats {
    ats!: AdvanceTacticsValues;
    vsAts!: AdvanceTacticsValues;

    constructor(statsSums: TacticStatsSums) {
        this.calculateAts(statsSums);
    }

    private calculateAts({ all, g, d, m, a }: TacticStatsSums): void {
        const dmCount = d.count + m.count;
        const dmTacklingAv = (d.tackling.sum + m.tackling.sum) / dmCount;
        const dmStrenghtAv = (d.strength.sum + m.strength.sum) / dmCount;
        const maCount = m.count + a.count;
        const maStrengthAv = (m.strength.sum + a.strength.sum) / maCount;

        const ats = {
            offside: [d.positioning.avg, d.speed.avg],
            pressingLow: [all.tackling.avg, all.speed.avg],
            pressingHigh: [all.passing.avg, all.speed.avg],
            counterAttack: [all.passing.avg, all.speed.avg],
            highBalls: [all.heading.avg, all.strength.avg],
            // Ats with one condition
            markingZonal: [dmTacklingAv, (d.speed.sum + m.speed.sum) / dmCount],
            markingMan: [dmTacklingAv, dmStrenghtAv],
            oneOnOnes: [(m.technique.sum + a.technique.sum) / maCount, maStrengthAv],
            gkStand: [g.reflexes.sum, g.handling.sum],
            gkRush: [g.agility.sum, g.outOfArea.sum],
            longShots: [(m.finishing.sum + a.finishing.sum) / maCount, (m.technique.sum + a.technique.sum) / maCount],
            firstTimeShots: [a.finishing.avg, a.heading.avg],
        };

        this.countAtWithOneConditionAverages(ats);
        this.ats = ats;

        const maPositioningAv = (m.positioning.sum + a.positioning.sum) / maCount;

        const vsAts = {
            offside: [a.positioning.avg, a.speed.avg],
            pressingLow: [all.tackling.avg, all.speed.avg],
            pressingHigh: [all.passing.avg, all.speed.avg],
            counterAttack: [all.passing.avg, all.speed.avg],
            highBalls: [all.heading.avg, all.strength.avg],
            // Ats with one condition
            markingZonal: [maPositioningAv, (m.speed.sum + a.speed.sum) / maCount],
            markingMan: [maPositioningAv, maStrengthAv],
            oneOnOnes: [dmTacklingAv, dmStrenghtAv],
            gkStand: [a.finishing.avg, a.heading.avg],
            gkRush: [a.technique.avg, a.heading.avg],
            longShots: [g.agility.avg, d.positioning.avg],
            firstTimeShots: [g.reflexes.avg, d.heading.avg],
        };

        this.countAtWithOneConditionAverages(vsAts);
        this.vsAts = vsAts;
    }

    private countAtWithOneConditionAverages(ats: AdvanceTacticsValues) {
        this.countAtAverage(ats.markingZonal);
        this.countAtAverage(ats.markingMan);
        this.countAtAverage(ats.oneOnOnes);
        this.countAtAverage(ats.gkStand);
        this.countAtAverage(ats.gkRush);
        this.countAtAverage(ats.longShots);
        this.countAtAverage(ats.firstTimeShots);
    }

    private countAtAverage(at: (number | undefined)[]): void {
        if (at[0] != undefined && at[1] != undefined) {
            at[2] = (at[0] + at[1]) / 2;
        } else {
            at[2] = 0;
        }
    }
}
