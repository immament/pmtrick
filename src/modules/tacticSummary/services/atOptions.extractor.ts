import log from 'loglevel';
import { singleton } from 'tsyringe';

import { AdvanceTacticTypes } from '../model/tacticStats.model';

type AdvAnceTacticSelectIdsWithValue =
    | 'Offside_Trap_1'
    | 'Pressing_1'
    | 'Pressing_2'
    | 'Counter_Attack_1'
    | 'High_Balls_1'
    | 'One_on_Ones_1'
    | 'Keeping_Style_1'
    | 'Keeping_Style_2'
    | 'Marking_1'
    | 'Marking_2'
    | 'RematesLonge_1'
    | 'RematesPrimeira_1';

const atsConvertMap = new Map<AdvAnceTacticSelectIdsWithValue, AdvanceTacticTypes>([
    ['Offside_Trap_1', 'offside'],
    ['Pressing_1', 'pressingHigh'],
    ['Pressing_2', 'pressingLow'],
    ['Counter_Attack_1', 'counterAttack'],
    ['High_Balls_1', 'highBalls'],
    ['One_on_Ones_1', 'oneOnOnes'],
    ['Keeping_Style_1', 'gkStand'],
    ['Keeping_Style_2', 'gkRush'],
    ['Marking_1', 'markingMan'],
    ['Marking_2', 'markingZonal'],
    ['RematesLonge_1', 'longShots'],
    ['RematesPrimeira_1', 'firstTimeShots'],
]);

@singleton()
export class AtOptionsExtractor {
    extract(): AdvanceTacticTypes[] {
        const atsElements = document.querySelectorAll<HTMLSelectElement>('#adv_options select');
        const ats = Array.from(atsElements).reduce((atsResult, at) => {
            const atResult = atsConvertMap.get(`${at.id}_${at.value}` as AdvAnceTacticSelectIdsWithValue);
            if (atResult) {
                atsResult.push(atResult);
            }
            return atsResult;
        }, [] as AdvanceTacticTypes[]);
        return ats;
    }
}

// type AdvanceTacticSelectIds =
//     | 'Offside_Trap'
//     | 'Pressing'
//     | 'Counter_Attack'
//     | 'High_Balls'
//     | 'One_on_Ones'
//     | 'Keeping_Style'
//     | 'Marking'
//     | 'RematesLonge'
//     | 'RematesPrimeira';
