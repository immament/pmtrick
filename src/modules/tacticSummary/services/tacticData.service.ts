import { inject, singleton } from 'tsyringe';

import { DataService } from '@src/common/services/data.service';
import { TacticEditorData } from '@src/modules/tacticSummary/model/tacticEditorPlayer.model';
import { AdvanceTacticTypes } from '@src/modules/tacticSummary/model/tacticStats.model';

@singleton()
export class TacticDataService {
    constructor(
        @inject('DataService<TacticEditorData>') public positions: DataService<TacticEditorData>,
        @inject('DataService<AdvanceTacticTypes[]>') public ats: DataService<AdvanceTacticTypes[]>,
    ) {}
}
