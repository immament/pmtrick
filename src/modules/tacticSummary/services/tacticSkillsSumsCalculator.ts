import { injectable } from 'tsyringe';

import { PlayerWithSkillsSummaries, SkillsSummaryCombo } from '@src/common/model/player.model';
import { PlayerSkillsEnum, PlayerSkillsType } from '@src/common/model/playerSkills.model';
import { SkillCalculatorService } from '@src/common/services/skillCalculator.service';
import { TacticEditorData, TacticEditorPosition } from '@src/modules/tacticSummary/model/tacticEditorPlayer.model';

import { StatRecord, StatsKeysArr, StatsPosition, TacticStatsSums } from '../model/tacticStats.model';

@injectable()
export class TacticSkillsSumsCalculator {
    private statsResult: TacticStatsSums = this.initStatsResult();

    constructor(private readonly skillCaluclatorService: SkillCalculatorService) {}

    calculate(data: TacticEditorData): TacticStatsSums | undefined {
        if (!data.positions?.length) return;
        this.statsResult = this.initStatsResult();
        this.calculateGs(data);

        this.calculateForPositions(data);
        this.calculateSumAll();

        this.calculateAvg();
        return this.statsResult;
    }

    private initStatsResult(): TacticStatsSums {
        const result = {} as TacticStatsSums;
        for (const position in StatsPosition) {
            const statRecord = { gs: { sum: 0 }, count: 0 } as StatRecord;
            result[position as StatsPosition] = Object.keys(PlayerSkillsEnum).reduce((acc, skill) => {
                acc[skill as PlayerSkillsEnum] = { sum: 0 };
                return acc;
            }, statRecord);
        }

        return result;
    }

    private calculateGs(data: TacticEditorData<PlayerWithSkillsSummaries>) {
        let index = 0;
        for (const position of data.positions) {
            if (++index > 11) break;
            if (!position.player) continue;
            const skillsSummaries: SkillsSummaryCombo = {
                current: this.skillCaluclatorService.calculateGs(position.player),
            };
            position.player.skillsSummaries = skillsSummaries;
            if (!position) continue;
        }
    }

    private calculateForPositions(data: TacticEditorData): void {
        let index = 0;
        for (const position of data.positions) {
            if (++index > 11) break;
            if (position.positionGroup) {
                this.statsResult[position.positionGroup].count++;
                this.statsResult.all.count++;
            }

            this.caluclateSkillsStats(position);
            this.calculateGsStat(position);
        }
    }

    private calculateSumAll() {
        const stats = this.statsResult;
        for (const skill in PlayerSkillsEnum) {
            const skillEnum = skill as PlayerSkillsEnum;
            stats.all[skillEnum].sum =
                stats.g[skillEnum].sum + stats.d[skillEnum].sum + stats.m[skillEnum].sum + stats.a[skillEnum].sum;
        }
        stats.all.gs.sum = stats.g.gs.sum + stats.d.gs.sum + stats.m.gs.sum + stats.a.gs.sum;
    }

    private roundAvg(n: number) {
        return Math.round((n + Number.EPSILON) * 100) / 100;
    }

    private calculateAvg() {
        for (const positionStat of Object.values(this.statsResult)) {
            for (const position of StatsKeysArr) {
                const skillStat = positionStat[position];
                skillStat.avg = positionStat.count > 0 ? this.roundAvg(skillStat.sum / positionStat.count) : undefined;
            }
        }
    }

    private caluclateSkillsStats(position: TacticEditorPosition): void {
        for (const skill in PlayerSkillsEnum) {
            this.caluclateStat(position, skill as PlayerSkillsType);
        }
    }

    private calculateGsStat({ player, positionGroup }: TacticEditorPosition<PlayerWithSkillsSummaries>) {
        if (player && positionGroup) {
            const skillValue = player?.skillsSummaries?.current?.best?.gs || 0;
            const currentStat = this.statsResult[positionGroup].gs;
            currentStat.sum = skillValue + currentStat.sum;
        }
    }

    private caluclateStat(position: TacticEditorPosition, skill: PlayerSkillsType): void {
        if (position.player && position.positionGroup) {
            const skillValue = position.player.skills[skill] || 0;
            const currentStat = this.statsResult[position.positionGroup][skill];
            currentStat.sum = skillValue + currentStat.sum;
        }
    }
}
