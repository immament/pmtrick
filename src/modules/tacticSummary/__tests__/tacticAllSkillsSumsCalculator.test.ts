import 'reflect-metadata';

import { container } from 'tsyringe';

import { GsFormula, tempGsFormula } from '@src/common/model/gsFormula.model';
import { TacticEditorData, TacticEditorPlayer } from '@src/modules/tacticSummary/model/tacticEditorPlayer.model';

import { TacticSkillsSumsCalculator } from '../services/tacticSkillsSumsCalculator';

describe('TacticAllSkillsSumCalculator', () => {
    beforeAll(() => {
        container.registerInstance<GsFormula>('GsFormula', tempGsFormula);
    });
    it('should create', () => {
        const calculator = container.resolve(TacticSkillsSumsCalculator);

        expect(calculator).toBeTruthy();
    });

    it('should calculate sum for one stat for all positions', () => {
        const calculator = container.resolve(TacticSkillsSumsCalculator);

        const result = calculator.calculate(mockData);
        expect(result).toMatchSnapshot();

        // expect(result).toStrictEqual(expectedResult);
    });
});

const mockData: TacticEditorData = {
    positions: [
        {
            id: '1000',
            position: 'gk',
            positionGroup: 'g',
            player: { skills: { speed: 11, strength: 6 } } as TacticEditorPlayer,
        },
        {
            id: '1001',
            position: 'dc1',
            positionGroup: 'd',
            player: { skills: { speed: 5, strength: 20 } } as TacticEditorPlayer,
        },
        {
            id: '1002',
            position: 'md',
            positionGroup: 'm',
            player: { skills: { speed: 11, strength: 0 } } as TacticEditorPlayer,
        },
        {
            id: '1002',
            position: 'mc1',
            positionGroup: 'm',
            player: { skills: { speed: 12, strength: 9 } } as TacticEditorPlayer,
        },
        {
            id: '1002',
            position: 'mc3',
            positionGroup: 'm',
            player: { skills: { speed: 20, strength: 1 } } as TacticEditorPlayer,
        },
        {
            id: '1003',
            position: 'a',
            positionGroup: 'a',
            player: { skills: { speed: 0, strength: 2 } } as TacticEditorPlayer,
        },
    ],
};
