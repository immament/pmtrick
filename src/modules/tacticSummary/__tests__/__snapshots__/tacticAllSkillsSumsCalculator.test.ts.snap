// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TacticAllSkillsSumCalculator should calculate sum for one stat for all positions 1`] = `
Object {
  "a": Object {
    "agility": Object {
      "avg": 0,
      "sum": 0,
    },
    "count": 1,
    "finishing": Object {
      "avg": 0,
      "sum": 0,
    },
    "gs": Object {
      "avg": 0,
      "sum": 0,
    },
    "handling": Object {
      "avg": 0,
      "sum": 0,
    },
    "heading": Object {
      "avg": 0,
      "sum": 0,
    },
    "outOfArea": Object {
      "avg": 0,
      "sum": 0,
    },
    "passing": Object {
      "avg": 0,
      "sum": 0,
    },
    "positioning": Object {
      "avg": 0,
      "sum": 0,
    },
    "reflexes": Object {
      "avg": 0,
      "sum": 0,
    },
    "speed": Object {
      "avg": 0,
      "sum": 0,
    },
    "strength": Object {
      "avg": 2,
      "sum": 2,
    },
    "tackling": Object {
      "avg": 0,
      "sum": 0,
    },
    "technique": Object {
      "avg": 0,
      "sum": 0,
    },
  },
  "all": Object {
    "agility": Object {
      "avg": 0,
      "sum": 0,
    },
    "count": 6,
    "finishing": Object {
      "avg": 0,
      "sum": 0,
    },
    "gs": Object {
      "avg": 0,
      "sum": 0,
    },
    "handling": Object {
      "avg": 0,
      "sum": 0,
    },
    "heading": Object {
      "avg": 0,
      "sum": 0,
    },
    "outOfArea": Object {
      "avg": 0,
      "sum": 0,
    },
    "passing": Object {
      "avg": 0,
      "sum": 0,
    },
    "positioning": Object {
      "avg": 0,
      "sum": 0,
    },
    "reflexes": Object {
      "avg": 0,
      "sum": 0,
    },
    "speed": Object {
      "avg": 9.83,
      "sum": 59,
    },
    "strength": Object {
      "avg": 6.33,
      "sum": 38,
    },
    "tackling": Object {
      "avg": 0,
      "sum": 0,
    },
    "technique": Object {
      "avg": 0,
      "sum": 0,
    },
  },
  "d": Object {
    "agility": Object {
      "avg": 0,
      "sum": 0,
    },
    "count": 1,
    "finishing": Object {
      "avg": 0,
      "sum": 0,
    },
    "gs": Object {
      "avg": 0,
      "sum": 0,
    },
    "handling": Object {
      "avg": 0,
      "sum": 0,
    },
    "heading": Object {
      "avg": 0,
      "sum": 0,
    },
    "outOfArea": Object {
      "avg": 0,
      "sum": 0,
    },
    "passing": Object {
      "avg": 0,
      "sum": 0,
    },
    "positioning": Object {
      "avg": 0,
      "sum": 0,
    },
    "reflexes": Object {
      "avg": 0,
      "sum": 0,
    },
    "speed": Object {
      "avg": 5,
      "sum": 5,
    },
    "strength": Object {
      "avg": 20,
      "sum": 20,
    },
    "tackling": Object {
      "avg": 0,
      "sum": 0,
    },
    "technique": Object {
      "avg": 0,
      "sum": 0,
    },
  },
  "g": Object {
    "agility": Object {
      "avg": 0,
      "sum": 0,
    },
    "count": 1,
    "finishing": Object {
      "avg": 0,
      "sum": 0,
    },
    "gs": Object {
      "avg": 0,
      "sum": 0,
    },
    "handling": Object {
      "avg": 0,
      "sum": 0,
    },
    "heading": Object {
      "avg": 0,
      "sum": 0,
    },
    "outOfArea": Object {
      "avg": 0,
      "sum": 0,
    },
    "passing": Object {
      "avg": 0,
      "sum": 0,
    },
    "positioning": Object {
      "avg": 0,
      "sum": 0,
    },
    "reflexes": Object {
      "avg": 0,
      "sum": 0,
    },
    "speed": Object {
      "avg": 11,
      "sum": 11,
    },
    "strength": Object {
      "avg": 6,
      "sum": 6,
    },
    "tackling": Object {
      "avg": 0,
      "sum": 0,
    },
    "technique": Object {
      "avg": 0,
      "sum": 0,
    },
  },
  "m": Object {
    "agility": Object {
      "avg": 0,
      "sum": 0,
    },
    "count": 3,
    "finishing": Object {
      "avg": 0,
      "sum": 0,
    },
    "gs": Object {
      "avg": 0,
      "sum": 0,
    },
    "handling": Object {
      "avg": 0,
      "sum": 0,
    },
    "heading": Object {
      "avg": 0,
      "sum": 0,
    },
    "outOfArea": Object {
      "avg": 0,
      "sum": 0,
    },
    "passing": Object {
      "avg": 0,
      "sum": 0,
    },
    "positioning": Object {
      "avg": 0,
      "sum": 0,
    },
    "reflexes": Object {
      "avg": 0,
      "sum": 0,
    },
    "speed": Object {
      "avg": 14.33,
      "sum": 43,
    },
    "strength": Object {
      "avg": 3.33,
      "sum": 10,
    },
    "tackling": Object {
      "avg": 0,
      "sum": 0,
    },
    "technique": Object {
      "avg": 0,
      "sum": 0,
    },
  },
}
`;
